### PR TITLE
"restaurants" collection and get most popular items functions

### PR DESCRIPTION
**Pull request by chdero**
Added 2 new google cloud functions:
**1** - **menuItemRated**: on menuitem document create, update (or create) the restaurant document that item belongs to, and add it to the list of rated items in reverse sorted order based on avgRating.

**2** - **menuItemUpdated**: on menuitem document update, update the restaurant document that item belongs to, change the avgRating for that menuitem in the restaurant docs list of rated items, and move the item up or down to maintain reverse sorted order based on avgRating.

Added 2 new functions to fsmanager.dart:
**1** - **getTopN(String restId, num numPopular)**: get the top 'nupPopular' rated items from restaurant with id 'restId'. Returns the list of maps, where each map is a menuItemId and the avgRating of that item. If there is less than 'numPopular' items rated for this restaurant, the function will return all the rated items.

**2** - **getTopFive(String restId)**: wrapper for 'getTopN'. This function returns the top five items (or less than 5 if there is less than 5 items rated) for restaurant with id 'restId'. Returns this in the same list of maps format described for 'getTopN'.

----

Tested the cloud functions in browser with firebase emulators:start, and tested the fsmanager functions with the backend test page.

----

Will deploy google cloud functions when this is accepted and merged.

### CODE REVIEW

File: firebase/functions/index.js [hunk 1]
```diff
@@ -198,4 +199,149 @@ exports.handleUserInfoDelete = functions.firestore
...
+
+                if(ratedItemList == null) {
+                    throw "Menu item updated, bet restaurant doc for item dne";
+                }
+
+                const listLen = ratedItemList.length;
+                var index = 0;
+                var i = 0;
+                // TODO: make more efficient (?)
+                for(index = 0; i < listLen; index++)
```
Comment (carteralbrecht) [inline firebase/functions/index.js:303]: is the i here intentional?

File: lib/services/fsmanager.dart [hunk 2]
```diff
@@ -51,6 +52,42 @@ class FirestoreManager {
...
+    try {
+      // Retrieve top items for a restaurant
+      dynamic topMenuItems =
+          await getDocData(restaurantCollection, restId, "ratedItems");
+      // Make sure the proper list was returned
+      if (topMenuItems is List) {
+        // Trim the list to hold only 'numPopular' items if necessary,
+        // and return the list
+        if (topMenuItems.length > numPopular) {
+          List<dynamic> topItems = new List(5);
```
Comment (carteralbrecht) [inline lib/services/fsmanager.dart:65]: should this be of variable length (not 5)

**Review by carteralbrecht — CHANGES_REQUESTED** [review]

File: lib/services/fsmanager.dart (hunk 2 above)
Comment (chdero) [reply lib/services/fsmanager.dart:65]: Yes it should be, I forgot to switch it over when I changed it from topFive to topN and made topFive call topN. Will fix it, ty.

File: firebase/functions/index.js (hunk 1 above)
Comment (chdero) [reply firebase/functions/index.js:303]: oops no. It still works cause of the break statement in the for loop, but ill fix it.

**Review by carteralbrecht — APPROVED** [review]
Different breed 😤

**Pull request merged**